mlx/linalg.h

```diff
@@ -0,0 +1,45 @@
+// Copyright © 2023 Apple Inc.
+
+#pragma once
+
+#include <variant>
+
+#include "array.h"
+#include "device.h"
+#include "ops.h"
+#include "stream.h"
+#include "string.h"
+
+namespace mlx::core::linalg {
+
+template <class... Ts>
+struct overloaded : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+
+/*
+Computes a vector norm.
+    If axes = {}, x will be flattened before the norm is computed.
+    Otherwise, the norm is computed over axes and the other dimensions are
+treated as batch dimensions.
+*/
+array vector_norm(
+    const array& a,
+    const std::variant<double, std::string>& ord = 2.0,
+    const std::vector<int>& axes = {},
+    bool keepdims = false,
+    StreamOrDevice s = {});
+array vector_norm(
+    const array& a,
+    const std::variant<double, std::string>& ord = 2.0,
+    bool keepdims = false,
+    StreamOrDevice s = {});
```

Comment:
I would prefer you do overloads for `const string& ord` and `double ord` rather than use a variant. There is no reason to do the dispatch at runtime when using the C++ API  so we should just have the overloads. I also think the variant makes the code a little cumbersome / harder to maintain.

Note you will have to update the python binding to do the dispatch instead.

Reply:
I will switch to overloads instead of the variant.